// @fleet-sdk/core/dist/esm/models/ergoUnsignedTransaction.js
import { utxoSum, utxoSumResultDiff } from "@fleet-sdk/common";
import { bytesToHex } from "@noble/hashes/utils";
import { serializeTransaction } from "../serializer/sigma/transactionSerializer";
import { blake2b256 } from "../serializer/utils";
export class ErgoUnsignedTransaction {
    _inputs;
    _dataInputs;
    _outputs;
    constructor(inputs, dataInputs, outputs) {
        this._inputs = Object.freeze(inputs);
        this._dataInputs = Object.freeze(dataInputs);
        this._outputs = Object.freeze(outputs);
    }
    get id() {
        return bytesToHex(blake2b256(this.toBytes()));
    }
    get inputs() {
        return this._inputs;
    }
    get dataInputs() {
        return this._dataInputs;
    }
    get outputs() {
        return this._outputs;
    }
    get burning() {
        return utxoSumResultDiff(utxoSum(this.inputs), utxoSum(this.outputs));
    }
    toPlainObject(outputType) {
        return {
            inputs: this.inputs.map((input) => input.toUnsignedInputObject(outputType || "default")),
            dataInputs: this.dataInputs.map((input) => input.toPlainObject(outputType || "default")),
            outputs: this.outputs.map((output) => _stringifyBoxAmounts(output))
        };
    }
    toEIP12Object() {
        return this.toPlainObject("EIP-12");
    }
    toBytes() {
        return serializeTransaction({
            inputs: this.inputs.map((input) => input.toUnsignedInputObject("default")),
            dataInputs: this.dataInputs.map((input) => input.toPlainObject("default")),
            outputs: this.outputs
        });
    }
}
function _stringifyBoxAmounts(output) {
    return {
        ...output,
        value: output.value.toString(),
        assets: output.assets.map((token) => ({
            tokenId: token.tokenId,
            amount: token.amount.toString()
        }))
    };
}
