// @fleet-sdk/core/dist/esm/models/ergoUnsignedInput.js
import { ErgoBox } from "./ergoBox";
export class ErgoUnsignedInput extends ErgoBox {
    _extension;
    get extension() {
        return this._extension;
    }
    constructor(box) {
        super(box);
        if (box.extension) {
            this.setContextVars(box.extension);
        }
    }
    setContextVars(extension) {
        this._extension = extension;
        return this;
    }
    toUnsignedInputObject(type) {
        return {
            ...this.toPlainObject(type),
            extension: this._extension || {}
        };
    }
    toPlainObject(type) {
        if (type === "EIP-12") {
            return {
                boxId: this.boxId,
                value: this.value.toString(),
                ergoTree: this.ergoTree,
                creationHeight: this.creationHeight,
                assets: this.assets.map((asset) => ({
                    tokenId: asset.tokenId,
                    amount: asset.amount.toString()
                })),
                additionalRegisters: this.additionalRegisters,
                transactionId: this.transactionId,
                index: this.index
            };
        }
        else {
            return {
                boxId: this.boxId
            };
        }
    }
}
