// @fleet-sdk/core/dist/cjs/models/ergoUnsignedTransaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErgoUnsignedTransaction = void 0;
const common_1 = require("@fleet-sdk/common");
const utils_1 = require("@noble/hashes/utils");
const transactionSerializer_1 = require("../serializer/sigma/transactionSerializer");
const utils_2 = require("../serializer/utils");
class ErgoUnsignedTransaction {
    _inputs;
    _dataInputs;
    _outputs;
    constructor(inputs, dataInputs, outputs) {
        this._inputs = Object.freeze(inputs);
        this._dataInputs = Object.freeze(dataInputs);
        this._outputs = Object.freeze(outputs);
    }
    get id() {
        return (0, utils_1.bytesToHex)((0, utils_2.blake2b256)(this.toBytes()));
    }
    get inputs() {
        return this._inputs;
    }
    get dataInputs() {
        return this._dataInputs;
    }
    get outputs() {
        return this._outputs;
    }
    get burning() {
        return (0, common_1.utxoSumResultDiff)((0, common_1.utxoSum)(this.inputs), (0, common_1.utxoSum)(this.outputs));
    }
    toPlainObject(outputType) {
        return {
            inputs: this.inputs.map((input) => input.toUnsignedInputObject(outputType || "default")),
            dataInputs: this.dataInputs.map((input) => input.toPlainObject(outputType || "default")),
            outputs: this.outputs.map((output) => _stringifyBoxAmounts(output))
        };
    }
    toEIP12Object() {
        return this.toPlainObject("EIP-12");
    }
    toBytes() {
        return (0, transactionSerializer_1.serializeTransaction)({
            inputs: this.inputs.map((input) => input.toUnsignedInputObject("default")),
            dataInputs: this.dataInputs.map((input) => input.toPlainObject("default")),
            outputs: this.outputs
        });
    }
}
exports.ErgoUnsignedTransaction = ErgoUnsignedTransaction;
function _stringifyBoxAmounts(output) {
    return {
        ...output,
        value: output.value.toString(),
        assets: output.assets.map((token) => ({
            tokenId: token.tokenId,
            amount: token.amount.toString()
        }))
    };
}
